### PR TITLE
Segwit: add buildStack

### PR DESCRIPTION
**Pull request by afk11**
Adds buildScript which avoids recursive methods which try to solve the input while trying to serialize. After expanding/signing you already know all of this, so we use some light code that decides where to put the results of buildScript.

It's kind of weird trying to do anything without making this assumption - the alternative is stuff like `var script=witnessScript||redeemScript||prevOutScript`

### CODE REVIEW

**Comment by afk11** [comment]
As mentioned, I can remove some code from that function once calling prepareInput guarantees that the input will be signable
 * that the txOut script is provided
 * if it's bare, it can only be [P2WPKH, P2WSH, P2SH, P2PKH, P2PK, Multisig]
 * if the txOut script calls for a P2SH, the redeemScript must be provided, and of [P2WPKH, P2WSH, P2PKH, P2PK, Multisig] ie, standard types and witness types. never p2sh.
 * if the txOut script calls for P2WSH, the witnessScript must be provided, and of [P2PKH, P2PK, Multisig] (ie, same types as standard-bare scriptPubKey)

I found changing expandOutput / prepareInput to accommodate this to be a real challenge (without the txOut), but would really suit the rest of signing.

File: src/transaction_builder.js [hunk 1]
```diff
@@ -462,15 +467,17 @@ TransactionBuilder.prototype.__build = function (allowIncomplete) {
...
     var result = buildInput(input, allowIncomplete)
 
     // skip if no result
-    if (!result && allowIncomplete) return
-    if (result && result.script) return tx.setInputScript(i, result.script)
-    if (result && result.witness) return tx.setWitness(i, result.witness)
+    if (!allowIncomplete) {
+      if (SIGNABLE.indexOf(result.type) === -1 && result.type !== bscript.types.P2WPKH) {
+        throw new Error(result.type + ' not supported')
+      }
```
Comment (dcousens) [inline src/transaction_builder.js]: Why do we need this?

File: src/transaction_builder.js [hunk 2]
```diff
@@ -224,72 +226,75 @@ function prepareInput (input, kpPubKey, redeemScript) {
...
-    case scriptTypes.MULTISIG:
-      signatures = signatures.map(function (signature) {
-        return signature || ops.OP_0
-      })
-
-      if (!allowIncomplete) {
-        // remove blank signatures
-        signatures = signatures.filter(function (x) { return x !== ops.OP_0 })
-      }
+function buildScript (type, signatures, pubKeys, allowIncomplete) {
```
Comment (dcousens) [inline src/transaction_builder.js]: `buildStack` more like

Comment (afk11) [inline src/transaction_builder.js]: +1 that would be more correct

File: src/transaction_builder.js (hunk 1 above)
Comment (afk11) [inline src/transaction_builder.js]: Normally we'd end up with [p2pkh, p2pk, multisig], but p2wpkh is just p2pkh, so when saving our knowledge about the inner-most script, we set the P2PKH solution (yep, solveOutput) for scriptCode if it's P2WPKH.

Comment (dcousens) [inline src/transaction_builder.js]: Man,  I know we both agree the complexity explosion in this area has always sucked, and made it so hard to review changes by any one.

Its all just so gross.

**Comment by afk11** [comment]
Bad rebase, easier to resubmit

**Pull request closed without merging**